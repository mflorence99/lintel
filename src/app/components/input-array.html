<ng-container
  [formGroup]="inputArrayForm">

  <div 
    [ngStyle]="{'grid-template-columns': 'repeat(auto-fit, ' + columnWidth + ')'}"
    class="inputs">

    <lintel-input
      [formControl]="input"
      *ngFor="let input of inputArrayForm.controls.inputs['controls']; let ix = index">

      <fa-icon 
        (click)="removeInput(ix)"
        [icon]="['fas', 'times']"
        *ngIf="inputArrayForm.controls.inputs['controls'].length > 1"
        suffix
        title="Remove this item">
      </fa-icon>

      <a
        (click)="addInput()"
        *ngIf="ix === (inputArrayForm.controls.inputs['controls'].length - 1) && ix < (maxItems - 1)"
        suffix>
        More
      </a>

    </lintel-input>

  </div>

</ng-container>
