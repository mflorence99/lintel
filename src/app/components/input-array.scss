:host {
  display: block;
}

.inputs {
  display: grid;
  grid-gap: 0.5rem;
  // NOTE: actual with set by style from control input
  // grid-template-columns: repeat(auto-fit, ...);
  overflow-y: auto;
}
