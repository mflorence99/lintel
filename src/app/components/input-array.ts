import { DestroyService } from '../services/destroy';

import { ChangeDetectionStrategy } from '@angular/core';
import { ChangeDetectorRef } from '@angular/core';
import { Component } from '@angular/core';
import { ControlValueAccessor } from '@angular/forms';
import { FormArray } from '@angular/forms';
import { FormBuilder } from '@angular/forms';
import { FormControl } from '@angular/forms';
import { FormGroup } from '@angular/forms';
import { Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { OnInit } from '@angular/core';

import { filter } from 'rxjs/operators';
import { forwardRef } from '@angular/core';
import { takeUntil } from 'rxjs/operators';

export type InputArrayType =(number | string)[];

/**
 * Input array form control
 *
 * NOTE: just enough to be able to match VSCode as well as possible
 *
 * @see https://blog.thoughtram.io/angular/2016/07/27/custom-form-controls-in-angular-2.html
 * 
 * Reference is old, but still helpful
 */

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => InputArrayComponent),
      multi: true
    },
    DestroyService
  ],
  selector: 'lintel-input-array',
  templateUrl: 'input-array.html',
  styleUrls: ['input-array.scss']
})

export class InputArrayComponent implements ControlValueAccessor, OnInit {

  @Input() columnWidth = '10rem';

  @Input() defaults: InputArrayType;
  @Input() enabled = true;

  inputArrayForm: FormGroup;

  @Input() max: number;
  @Input() maxItems = Infinity;
  @Input() min: number;

  @Input() step: number;

  @Input() type: 'number' | 'text' = 'text';

  @Input() uniqueItems: boolean;

  values: InputArrayType = [];

  @Input()
  get value(): InputArrayType {
    return this.values;
  }
  set value(value: InputArrayType) {
    this.values = value ?? [];
    this.underConstruction = true;
    const inputs = this.inputArrayForm.controls.inputs as FormArray;
    // NOTE: there's always one more input for a new value
    // trim off excess inputs
    while (inputs.length > (this.values.length + 1))
      inputs.removeAt(inputs.length - 1);
    while (inputs.length < (this.values.length + 1))
      inputs.push(new FormControl(null));
    // patch in the new values
    inputs.patchValue([...this.values, null], { emitEvent: false });
    this.underConstruction = false;
    this.onChange?.(this.values);
    this.cdf.detectChanges();
  }

  private onChange: Function;
  private underConstruction: boolean;

  /** ctor */
  constructor(private cdf: ChangeDetectorRef,
              private destroy$: DestroyService,
              private formBuilder: FormBuilder) { 
    // initialize the form
    this.inputArrayForm = this.formBuilder.group({
      inputs: new FormArray([])
    });
  }

  /** Add another input */
  addInput(): void {
    if (this.enabled) {
      const inputs = this.inputArrayForm.controls.inputs as FormArray;
      inputs.push(new FormControl(null));
    }
  }

  /** When we're ready */
  ngOnInit(): void {
    this.inputArrayForm.valueChanges
      .pipe(
        filter(_ => !this.underConstruction),
        takeUntil(this.destroy$)
      )
      .subscribe(value => {
        this.values = value.inputs
          .filter(val => !!val)
          .map(val => (this.type === 'number') ? Number(val) : val);
        if (this.uniqueItems)
          this.values = Array.from(new Set(this.values));
        this.onChange?.(this.value);
      });
  }

  /** @see ControlValueAccessor */
  registerOnChange(fn): void {
    this.onChange = fn;
  }

  /** @see ControlValueAccessor */
  registerOnTouched(_): void { }

  /** Remove specified input */
  removeInput(ix: number): void {
    if (this.enabled) {
      const inputs = this.inputArrayForm.controls.inputs as FormArray;
      inputs.removeAt(ix);
    }
  }

  /** @see ControlValueAccessor */
  writeValue(value): void {
    this.value = value;
  }

}
