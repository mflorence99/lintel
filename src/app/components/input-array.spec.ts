import { InputArrayComponent } from './input-array';
import { TestBed } from '@angular/core/testing';

import { async } from '@angular/core/testing';
import { prepare } from './component.spec';

describe('InputArrayComponent', () => {

  beforeEach(async(() => prepare()));

  test('Component is created', () => {
    const fixture = TestBed.createComponent(InputArrayComponent);
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });

});
