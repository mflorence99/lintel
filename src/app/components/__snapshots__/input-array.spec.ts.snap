// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputArrayComponent snapshot 1`] = `
<lintel-input-array
  __ngContext__={[Function LRootView]}
  cdf={[Function ViewRef]}
  columnWidth={[Function String]}
  destroy$={[Function DestroyService]}
  enabled={[Function Boolean]}
  formBuilder={[Function FormBuilder]}
  inputArrayForm={[Function FormGroup]}
  maxItems={[Function Number]}
  type={[Function String]}
  underConstruction="false"
  uniqueItems={[Function Boolean]}
  values={[Function Array]}
>
  <p
    class="inputs"
    style="grid-template-columns: repeat(auto-fit, 10rem);"
  >
    
    <lintel-input
      class="ng-untouched ng-pristine ng-valid"
    >
      <div
        prefix=""
      />
      
      <input
        autocomplete="off"
        max="undefined"
        min="undefined"
        placeholder=""
        step="undefined"
        tabindex="0"
        type="text"
      />
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      
      
    </lintel-input>
    <lintel-input
      class="ng-untouched ng-pristine ng-valid"
    >
      
      <input
        autocomplete="off"
        max="undefined"
        min="undefined"
        placeholder=""
        step="undefined"
        tabindex="0"
        type="text"
      />
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      
      
    </lintel-input>
    <lintel-input
      class="ng-untouched ng-pristine ng-valid"
    >
      
      <input
        autocomplete="off"
        max="undefined"
        min="undefined"
        placeholder=""
        step="undefined"
        tabindex="0"
        type="text"
      />
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      
      
    </lintel-input>
    <lintel-input
      class="ng-untouched ng-pristine ng-valid"
    >
      
      <input
        autocomplete="off"
        max="undefined"
        min="undefined"
        placeholder=""
        step="undefined"
        tabindex="0"
        type="text"
      />
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      
      
    </lintel-input>
    <lintel-input
      class="ng-untouched ng-pristine ng-valid"
    >
      
      <input
        autocomplete="off"
        max="undefined"
        min="undefined"
        placeholder=""
        step="undefined"
        tabindex="0"
        type="text"
      />
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      <a
        suffix=""
      >
         More 
      </a>
      
      <div
        suffix=""
      />
      
    </lintel-input>
    
  </p>
</lintel-input-array>
`;
